python/ray/air/_internal/checkpoint_manager.py

```diff
@@ -286,6 +286,8 @@ def __init__(
 
         # Used for keeping top K checkpoints.
         self._top_persisted_checkpoints: List[_HeapCheckpointWrapper] = []
+        # Also keep a set of all existing checkpoints
+        self._all_persisted_checkpoint_data: Set[Hashable] = set()
```

Comment:
The main problem is in function trainables. When I schedule multiple `trainable.save.remote()` calls they will all register a separate checkpoint with the same checkpoint dir (the step number). This means effectively we'll keep less checkpoints than `num_to_keep`.

Previously this wasn't a problem because `save_to_object` used a temporary checkpoint directory